pybop/observers/unscented_kalman.py

```diff
@@ -0,0 +1,309 @@
+            + [x + eta * S[:, i].reshape(-1, 1) for i in range(L)]
+            + [x - eta * S[:, i].reshape(-1, 1) for i in range(L)]
+        )
+        w_m = np.array([wm_0] + [(1 - wm_0) / (2 * L)] * (2 * L))
+        w_c = np.array([wc_0] + [(1 - wc_0) / (2 * L)] * (2 * L))
+        return (points, w_m, w_c)
+
+    @staticmethod
+    def unscented_transform(
+        sigma_points: np.ndarray, w_m: np.ndarray, w_c: np.ndarray, sqrtR: np.ndarray
+    ) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Performs the unscented transform
+
+        Parameters
+        ----------
+        sigma_points : List[SigmaPoint]
+            The sigma points
+        sqrtR : np.ndarray
+            The square root of the covariance matrix
+
+        Returns
+        -------
+        Tuple[np.ndarray, np.ndarray]
+            The mean and covariance of the sigma points
+        """
+        x = np.sum(w_m * sigma_points, axis=1).reshape(-1, 1)
+        A = np.sqrt(w_c[1:]) * (sigma_points[:, 1:] - x)
+        (
+            _,
+            S,
+        ) = linalg.qr(A.T, mode="economic")
+        S = UkfFilter.cholupdate(S, sigma_points[:, 0:1] - x, w_c[0])
+        return x, S
+
+    @staticmethod
+    def cholupdate(R: np.ndarray, x: np.ndarray, w: float) -> np.ndarray:
+        """
+        Updates the cholesky decomposition of a matrix (see https://github.com/modusdatascience/choldate/blob/master/choldate/_choldate.pyx)
+
+        Note: will be in scipy soon so replace with this: https://github.com/scipy/scipy/pull/16499
+
+        TODO: need to replace with something more low-level
+
+        Parameters
+        ----------
+        R : np.ndarray
+            The cholesky decomposition of the matrix
+        x : np.ndarray
+            The vector to add to the matrix
+        w : float
+            The weight of the vector
+
+        Returns
+        -------
+        np.ndarray
+            The updated cholesky decomposition
+        """
+        x = x.flatten()
+        p = x.shape[0]
+        for k in range(p):
+            r = np.sqrt(R[k, k] ** 2 + w * x[k] ** 2)
+            # r = UkfFilter.hypot(R[k, k], x[k])
+            c = r / R[k, k]
+            s = x[k] / R[k, k]
+            R[k, k] = r
+            if k < p - 1:
+                R[k, k + 1 :] = (R[k, k + 1 :] + w * s * x[k + 1 :]) / c
+                x[k + 1 :] = c * x[k + 1 :] - s * R[k, k + 1 :]
+        return R
+
+    def step(self, y: np.ndarray) -> float:
+        """
+        Steps the filter forward one step using a measurement. Returns the log likelihood of the measurement.
+
+        Parameters
+        ----------
+        y : np.ndarray
+            The measurement vector
+
+        Returns
+        -------
+        float
+            The log likelihood of the measurement
+        """
+        sigma_points, w_m, w_c = self.gen_sigma_points(
+            self.x, self.S, self.alpha, self.beta
+        )
+        sigma_points = np.apply_along_axis(self.f, 0, sigma_points)
```

Comment:
I wonder if we can vectorise `self.f` and `self.h` to avoid using `np.apply_along_axis`? https://stackoverflow.com/questions/23849097/numpy-np-apply-along-axis-function-speed-up

Reply:
Yea, this would have to be done somewhere in PyBaMM, needs to be a way to solve many identical pybamm models all starting from different initial conditions.

Reply:
see https://github.com/pybamm-team/PyBaMM/issues/3713